File: Shreya/Tree/Path Sum II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void findpath(TreeNode* root, int targetSum, vector<int>&curr, vector<vector<int>> &paths)
    {
        if (root==NULL)
        {
            return;
        }
        curr.push_back(root->val);
        if ((targetSum-root->val==0)&&(root->left==NULL)&&(root->right==NULL))
        {
            paths.push_back(curr);
            curr.pop_back();
            return;
        }
        findpath(root->left,targetSum-root->val,curr,paths);
        findpath(root->right,targetSum-root->val,curr,paths);
        curr.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
       vector<vector<int>>paths;
       vector<int>curr;
       findpath(root,targetSum,curr,paths);
       return paths; 
    }
};

File: Rohit Singhal/Trees/PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> ans=new ArrayList<>();
    List<Integer> temp=new ArrayList<>();
    
    public void dfs(TreeNode root, int targetSum)
    {
        if(root==null)
            return;
        
         if(root.left==null && root.right==null)
        {
            if(targetSum==root.val)
            {
                temp.add(root.val);
                ans.add(new ArrayList<>(temp));
                temp.remove(temp.size()-1);
                return;
            }
             else
             {
                 return;
             }
        }
        temp.add(root.val);
        dfs(root.left,targetSum-root.val);
        dfs(root.right,targetSum-root.val);
        temp.remove(temp.size()-1);
        return;
        
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        dfs(root,targetSum);
        return ans;
       
    }
}
